csharp/src/SeedLang.Block/Position.cs

```diff
@@ -0,0 +1,60 @@
+// Copyright 2021 The Aha001 Team.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using SeedLang.Common;
+
+namespace SeedLang.Block {
+  // The position where a block docks to another block, or stays as a standalone block on a
+  // SeedBlock module canvas.
+  //
+  // This class is different with SeedLang.Common.BlockPosition, which represents a syntaxical code
+  // position in a SeedBlock module.
+  public class Position {
+    // How the block is docked to another block.
+    public enum DockType {
+      // The block is standalone, not docked to any other block. The block's position is defined by
+      // CanvasPosition.
+      UnDocked = 0,
+      // The block docks as one of the target block's input blocks. The index of the input position
+      // is specified by DockPosition.
+      Input,
+      // The block docks as the next statement of the target block. In this case, DockPosition is
+      // omitted.
+      NextStatement,
+      // The block docks as the first child statement at one of the target block's dock slots. The
+      // index of the dock slot is specified by DockPosition.
+      ChildStatement,
+    }
+
+    public DockType Type { get; set; }
+    public Vector2 CanvasPosition = new Vector2(0, 0);
+    public string TargetBlockId { get; set; }
+    public int DockPosition { get; set; }
+    public bool IsDocked => Type != DockType.UnDocked;
+
+    // Initializes an un-docked position. It's canvas position is (x, y).
+    public Position(int x, int y) {
```

Comment:
I tried but it seems pretty messy if we defined them as two classes. Consider:

We still need a base class Position since a block can be either docked to another block or, located at a specified canvas position. So the hierarchy of the 3 classes might be:

```
class Position {
  enum DockType {...}
  DockType type;
  bool IsDocked { ... }
}

class CanvasPosition : Position {
  int X;
  int Y;
}

class DockPosition : Position {
  string targetBlockId;
  int DockSlotIndex;
}
```

Following the above design, there are many inconveniences. E.g.,

1) When we need to get the target block ID from a block's position, we need to cast its type to DockPosition:

```
string targetBlockId = (block.Pos as DockPosition).targetBlockId;
```

Or, when we simply want to get the canvas position from a block's position:

```
Vector2 point = (block.Pos as CanvasPosition).GetVector2();
```

2) When a block's position is changing from a canvas position to a dock position (or vice versa), we cannot simply update the fields of block.Pos. Instead, we need to drop the existing instance and assign a new one to block.Pos:

```
block.Pos = new DockPosition(type, targetBlockId, dockSlotIndex);
```

Any better ideas?

Reply:
I changed the property name DockPosition to DockSlotIndex in the Position class and updated the comments, to make the terms and the concepts clearer.

Reply:
My original idea is:

class Position {
    bool IsDocked { return this is DockPosition; }
}

class CanvasPosition: Position {
    int X; int Y;
}

class DockPosition: Position {
    enum DockType { no UnDocked type in it, only "input, statement, etc." }
    DockType Type;
    string TargetBlockId;
    int DockSlotIndex;
}

The Position class is just used to store two types. But in most cases the client knows which one it need use, and can cast to the concrete type.
I don't know if it will simplify the code or not, and there isn't big difference from your design. You can decide which one is better.

Reply:
No difference. By following your design, we still need to use explicit type casting everywhere to get or set the field values, like:

```
if (block.IsDocked) {
  (block as DockPosition).targetBlockId = someIdElse;
}
```

Let's keep the current *simple* solution for now.